### ISSUE TITLE
Read-ahead or Prefetch configurable values

### ISSUE THREAD

**Issue by lars18th** [nova-video-player/aos-AVP#420]
Hi,

I'm testing the use of the SFTP access, and it works! I'm happy, but I see one problem: when using it in a LAN the functionality is very rebust. However, when using it over a VPN (remote access, with a large network latency), then even if the play works, I see a very large starting time (between 30 to 60 seconds). Futhermore, I've checked that the problem is not the bandwith, as the "streaming" (the reading of the file over the SSH connection) is not using the full bandwith of the network. And the play is good, as the playing never stops because insufficient data. Then, I feel the problem is related to the _buffer size_ **and** the _read chunk size_. Regarding the last, it's easy to check that NVP is not reading at full speed.

So, I request if you can add the option to configure the SIZE of the _playing_ buffer, and the SIZE of the _reading_ buffer. The second is the one related to the read-ahead of next chunks (aka prefetching). So increasing it the delay of actions with huge network latency values will be decreased.

What you think?

**Comment by lars18th** [comment]
Hi,

As I don't see any response, I suggest for one idea to improve the _efficiency_ when using SFTP connections: Why not include a _LOCAL CACHE_ of files in the local storage? The idea is quite simple: when listing the SFTP share then you compare the content with the local cache. If some file is different, then remove it from the local cache. Then when loading a file, use the local storage _only if the full file is there_. And limit the size of files to small values (until 1024KB, for example). Futhermore, every time you write a file, a local copy is stored. This will decrease the response time in several orders of magnitude. And for reading changes in the metada (*.NFO files and thumbnails) it will be similar to local content.

After that, the next step will give a threaded reading of the playing files (MKV, TS, MP4, etc.) with some read-ahead buffer. As I can see in the current source, the reading is done synchrounsly, and for this reason the playing start is very slow.

Any feedback?

**Comment by phhusson** [comment]
Hi,

What's your latency, available bandwidth to sftp server, and media bandwidth?

Startup buffering is hard-coded at 24MB in https://github.com/nova-video-player/aos-AVP/wiki/Buffering

If you're playing low bandwidth stream on a 1-5Mbps upload server, that'll make ~ 1 minute startup time

**Comment by lars18th** [comment]
Hi,

> What's your latency, available bandwidth to sftp server, and media bandwidth?

Bandwith: 10-15Mbps max.; over SFTP between 6-12Mbps of real throughput.
Latency: >100ms and less than 500ms.

> Startup buffering is hard-coded at 24MB in https://github.com/nova-video-player/aos-AVP/wiki/Buffering

So, why not make this value configurable, almost with a file configuration (whitout changes in the user interface)?
I suggest, a value around 8MB.

> If you're playing low bandwidth stream on a 1-5Mbps upload server, that'll make ~ 1 minute startup time

Yes! Over WAN I see around 1 minute start time, and over LAN this is reduced to 15 seconds.
I feel there is some space for improve this. Perhaps the simple startup buffer configuration can be one initial step.

Regards.

**Comment by phhusson** [comment]
> So, why not make this value configurable, almost with a file configuration (whitout changes in the user interface)?

That's still additional work for a pretty limited use-case. (it's possible other people will have same issue, but the number of those people who would know to edit config file is pretty slim). Anyway, we're not totally sure about our understanding of our buffering, so could you recompile the APK with 8MB to confirm this is the issue?

> Yes! Over WAN I see around 1 minute start time, and over LAN this is reduced to 15 seconds.

Uh 15s over LAN is very high (unless it's a 20Mbps WiFi)

**Comment by courville** [comment]
If it is changed, we need an adaptive formula that will take into account real throughout bandwidth (how to probe it?), video and audio bitrate of the video file etc...

**Comment by lars18th** [comment]
> If it is changed, we need an adaptive formula that will take into account real throughout bandwidth (how to probe it?), video and audio bitrate of the video file etc...

That's in fact the real solution! You need to calculate the buffer based on time. For example you can use this strategy:
- Initial buffer size: 1MB.
  - Start to read, and determine the FRAME-RATE.
  - Based on the FRAME-RATE read 1/2 second; and if the buffer is full, then increase size in steps of 1MB chunks.
  - After readed the 1/2 second, determine the BITRATE; and based on the bitrate calculate the final buffer size based on user configured time (1/2 second, 1 second, 1.5 seconds, etc.).
  - If when playing the buffer is full, then assume the source is VBR and increase value.

Using this algorithm you can implement an adaptative reading buffer.
I hope this helps.

**Comment by lars18th** [comment]
> Anyway, we're not totally sure about our understanding of our buffering, so could you recompile the APK with 8MB to confirm this is the issue?

In my case, I need for a configurable value, as for some videos the current 24 is small. But for regular ~5Mbps VBR H.264 videos over SFTP I need to reduce it to a half or quarter of the current value.

**Comment by lars18th** [comment]
Hi,

Sorry for the stupid algorithm presented before. That's an enhanced idea:
- First, increase the buffer size to 64MB. This will improve the speed load over slow links.
- Second, get the FRAME-RATE of the file to play (that's mandatory).
- When the buffer has 1/2 second, or 1 second or 1.5 seconds (configurable by user) then START to PLAY.
- However, continue reading the buffer until it will full, and refill it when it has some empty space.

This strategy produces a sufficient look-ahead reading of the buffer, and starts to play fast. And futhermore, it works with any type of source (network share, local storage, SFTP, etc.) and doesn't disturbs the current play. You only need to rewrite the buffer handle.

I hope you want to implement this or something similar.
Thank you for this very good player. I hope this idea will help you.
Regards.

**Comment by phhusson** [comment]
We know we need to take video timestamps into account, and it's been in the todo list for years, but so far we've considered it's too much work.

**Comment by lars18th** [comment]
> We know we need to take video timestamps into account, and it's been in the todo list for years, but so far we've considered it's too much work.

Then, perhaps with some collaboration we can help. From my point of view, the first, and perhaps the simple step is to implement a callback function to instruct to the player to start to play. Then with some help we can try to implement the function that checks the tiemestamps and calculate when the buffer has sufficient data to start playing. Futhermore, with this "simple" enhancement can be possible to implement initially a dumb strategy: Start to play when the buffer is filled at XX % of size. This is trivial to implement, but it improves like a lookahead reading. For example, if you double the buffer size to 48MB and start to play at 24MB, the current behaviour from the user perspective will be near identical; but in fact you will look-ahead for a 24MB. All without much effort (just the start play callback and the new size of the buffer).

What you think?

**Comment by mickeul** [comment]
I vote 1 UP to make the possibility to increase buffer size (in MB or in milliseconds as VLC does)
I have UHD HDR10 videos in a local NAS and the video stops every 4 min
In Kodi, no, neither in VLC

**Comment by lars18th** [comment]
Hi @mickeul ,

Not only to increase the buffer size, but to reduce it too!
I suggest to do this soon, even using an XML configuration, that can be implemented for example at folder level with the `*.archos.xml` files.

However, the fundamental required functionality will be the enabling of the dynamic use of the buffer. This will be achieved when the start of playback is based on time rather than size.

I hope this will helps the developers to improve this good player.

**Comment by mickeul** [comment]
The goal is to make available the opportunity to change the buffer size (increase or decrease) all other vidéo players can do it so Nova should too
And has to be available easily, not only via a trivial modification like editing the xml file, it has to be well accessible

**Comment by phhusson** [comment]
Actually I was wrong, the line mentioned does NOT hard-code startup buffer size.

@lars18th did you try changing its value, and actually see improvement?

**Comment by lars18th** [comment]
> Actually I was wrong, the line mentioned does NOT hard-code startup buffer size.
> 
> @lars18th did you try changing its value, and actually see improvement?

Sorry @phhusson , at time I don't have tested it.

**Comment by Pentaphon** [comment]
> The goal is to make available the opportunity to change the buffer size (increase or decrease)

I agree with this. It seems like something that should be available to the user as an advanced option rather than be hardcoded. 

@courville do you think the ability to change buffer size is worth pursuing? I've been running into buffering issues on 4K video sources, even on my very rock solid network connection and I think Nova users will begin to come across these issues more as they start to play more 4K videos over their home network. 

This is also yet another reason that mpv may be worth using as a replacement for ffmpeg as all of these values are easily configurable by the user and has has their community come up with good default settings. 

https://mpv.io/manual/master/#cache

Then again, we could just switch to the default cache setting that Kodi uses, which is a maximum of 60MB of free RAM, much higher than the 24MB Nova currently uses. I just tried the same 4K video file on both Kodi and Nova on the same 4K Fire Stick using the same home network and I have to say, Nova buffers very slightly at times, while Kodi doesn't buffer at all.  

https://kodi.wiki/view/HOW-TO:Modify_the_video_cache

**Comment by Pentaphon** [comment]
@phhusson @courville I've been playing 4K files and I've been running into stuttering issues more than I'd like. I think the current buffering settings should be re-visited and re-adjusted for today's increasingly 4K-heavy video landscape. I don't think that it is totally necessary to let users adjust buffering settings themselves, but the defaults most definitely need to be increased for 4K playback.

**Comment by courville** [comment]
I will add a hidden option to play with buffering sizes for you to experiment and report if any change would be beneficial.

**Comment by Pentaphon** [comment]
> I will add a hidden option to play with buffering sizes for you to experiment and report if any change would be beneficial.

Thank you!

**Comment by courville** [comment]
I have a version that has max compressed iframe size and stream buffer size to play with.
Unfortunately, I do not think it will help.
What I can observe is that over SMB shares using Wi-Fi or Ethernet nova stutters on very large 4k videos while other players don't.
However this is not the case over SFTP.
For me the issue is more linked to https://github.com/AgNO3/jcifs-ng/issues/263
See https://github.com/nova-video-player/aos-AVP/issues/420 also.
Can you try if it helps to use SFTP?

**Comment by courville** [comment]
Noet to self: https://stackoverflow.com/questions/43109872/performance-android-jcifs-file-reading-is-slow
-> not relevant too old.

**Comment by Pentaphon** [comment]
> Can you try if it helps to use SFTP?

Unfortunately I don't use SFTP since it's simply not as quick to use as SMB which is very fast and simple to use on Windows or Linux. However, if you say that SFTP is faster on Nova at the movie then I will take your word for it and I might give it a shot if I can find tiny, simple SFTP solutions for Windows and Linux. 

>For me the issue is more linked to https://github.com/AgNO3/jcifs-ng/issues/263

If that issue with AgNO3 is actually the problem then I apologize for pressing this issue and we have to wait for that project to fix that issue in order to get good SMB performance. 

May I ask though, to get back to the topic: Do you think that Nova's current buffer values are sufficient for 4K playback over Wifi or do they need to be tweaked, regardless of which network protocol (SMB, FTP, DLNA) is being used to serve files?

**Comment by courville** [comment]
@Pentaphon, I will let you play wit the buffer tweaking but I do believe that the current settings are really ok.
Solution for increasing SMB performance would be either i) wait for mbichler to implement multi-credit requests but this could take a while (or never happen), ii) switch to https://github.com/hierynomus/smbj which is limited to SMB2/3 (users will complain about loss of SMB1) and would require some tedious work on my side.
Now if you want to play with SFTP it is natively supported on linux and you can enable it easily on windows too.
I agree that SMB is the easiest option of all for being really user friendly.

**Comment by Pentaphon** [comment]
> ill let you play wit the buffer tweaking but I do believe that the current settings are really ok.

Fair enough. I wasn't aware that the current SMB implementation was currently that slow so that may indeed be the actual core issue at hand. However, if buffer tweaking can help with the current slow SMB performance then perhaps it may help users see less buffering. 

> Solution for increasing SMB performance would be either i) wait for mbichler to implement multi-credit requests but this could take a while (or never happen), ii) switch to https://github.com/hierynomus/smbj which is limited to SMB2/3 (users will complain about loss of SMB1) and would require some tedious work on my side.

Well considering that 1) SMB1 is very insecure by today's standards 2) everybody should be using an up-to-date OS/NAS that supports SMB2/3 these days if they want their OS/NAS to be secure in general 3) the current SMB implementation is just too slow to be useful for 4K files and anything beyond that like 4K Remuxes which I've seen some people archive, I think moving to hierynomus/smbj is worth it in the long run, especially considering that SMB is the simplest way for everybody to share their media. 

>Now if you want to play with SFTP it is natively supported on linux and you can enable it easily on windows too.

I just tried playing with it on Windows and it's definitely not as easy to set up there. Also, SFTP shares, when added to the library, don't seem to get fully scraped by Nova. I'm opening an issue for that here: https://github.com/nova-video-player/aos-AVP/issues/744

**Comment by Pentaphon** [comment]
> Can you try if it helps to use SFTP?

@courville So I tried SFTP with a bunch of high-ish bitrate 4K files that would stutter over SMB and it's definitely faster. I just wish Nova could scrape the entire root share the same way Nova does with SMB shares. 

A configurable buffer value may still mitigate the current stuttering over SMB though. I'd have to experiment. I do think that I could find a better buffer value that may end up helping all users who play 4K files. Any change the "newlookv6" UI has readable buffer display functionality to help me figure this out? The current "read ahead" faded portion on the progress bar is nice but it's not quite enough for numerical value testing.

**Comment by Pentaphon** [comment]
@lars18th @mickeul the stream buffer size option was just added in the 6.0.86 build if you want to get started on tweaking it for your needs. Thanks to courville for making this happen.

**Comment by courville** [comment]
FYI in order to access the buffer sizes in nova settings you need to enable advanced option by clicking several times on enable sw decoding (6 or so). However it will no help with SMB because bottleneck is on jcifs-ng and buffer starvation will occur.

**Comment by Pentaphon** [comment]
@courville I was just playing around with the stream buffer size option and it has yielded a massive improvement in playback for me, especially for 4K files. There's no stuttering with higher bitrate files over Wifi anymore since I now have a visible readahead buffer in the progress bar with a healthy amount of buffer. Great work! I think a lot of people will appreciate this! I've had no problem raising the buffer from 24MB to 64MB, 100MB, 150MB, etc. The option is a huge help, especially on SFTP which has no bottleneck at the moment. 

However, there is one major issue left while torture-testing Nova's new configurable buffering and that's that something I've opened an issue for at https://github.com/nova-video-player/aos-AVP/issues/747

With this new setting, I think this issue can be closed. Thanks again!

>FYI in order to access the buffer sizes in nova settings you need to enable advanced option by clicking several times on enable sw decoding (6 or so).

Yeah I was able to figure that out on my streaming devices from other posts you've made but the option is readily available on the phone UI.

**Comment by courville** [comment]
@Pentaphon, thanks for the feedback.
Does it help in your case for smb (it should not)?

**Comment by Pentaphon** [comment]
> Does it help in your case for smb (it should not)?

I think it does, actually. The extra read-ahead gives me enough buffer data to not have any stutter even on SMB with the current bottleneck. Before, I would get stutter on 4K movies but now I have yet to experience that on the same movies on the same network from the same SMB shares. I think 24MB is actually too little for 4K content based on my testing. Have you tested the default value with 4K content? Hopefully other people reading this thread can do some testing and see if they got better streaming playback from tweaking their stream buffer size.

**Comment by Pentaphon** [comment]
Ok I've been stress testing the read-ahead buffer setting and it is the biggest improvement I've ever seen for Nova by far. 

On SMB using 4K files: increasing the buffer makes a huge difference in streaming quality since the buffer is actually allowed to go as far ahead as you want. 

On SFTP using 4K files: not much difference because SFTP seems faster on my Windows machine and 24 MB works for all. 

There's a huge difference on both on when using Wifi because even if you have an unreliable connection, the buffer can be set high enough so that even if your Wifi drops for a few seconds, the buffer keeps going. 

I've come to the conclusion that literally any increase in read-ahead buffer helps Nova and I've personally settled on 100MB as a good round number to use for my Nova install. It's over 4x the amount of default buffer, which you can see on the progress bar easily and it makes a massive difference for me. I'm never using the default setting again. 100MB is not too big, not too small, and it works even if something goes wrong with Wifi, giving you enough buffer to not worry about playback being interrupted. I've seen other players like mpv use a default cache of 150MB and it works for them so 100MB is not unreasonable, nor unheard of. 

I would advise 2 things as a humble user:

1) Tuck the read-ahead buffer setting into an "advanced settings" page. Don't hide it behind some "secret keypresses". If VLC can let people change this setting easily, so can Nova. 

2) Increase the default buffer from 24MB to 48MB. That's the lowest number where I started seeing an improvement in all media, from 4K to high-bitrate Bluray, etc. Most people will use SMB and this extra buffer will get rid of most if not all playback interruptions. 

Just my two cents. Thanks again, @courville

**Comment by Pentaphon** [comment]
@courville I believe we can close this with the new hidden configurable buffer values

I also think smbj may have to be a high priority after webdav support is integrated.
